### DISCUSSION TITLE
Need help understanding

### DISCUSSION THREAD

**Discussion by y9Kap** [terrorsl/sMQTTBroker#43] [Q&A] [answered]
Hi. I'm facing a common error MQTT_IDENTIFIER_REJECTED which causes my client to stop connecting to the broker. I went to look at the source and found this: https://github.com/terrorsl/sMQTTBroker/blob/db9689dabd4b25527bec5b87525511b384a63739/src/sMQTTBroker.cpp#L231
I don't understand, aren't these two checks contradictory? The first condition returns false if the clients match, and the second returns true if the clients match.

**Comment by terrorsl** [comment]
Hello, this function is called when the client is already in the client list and has passed its identifier. It checks that its identifier is unique and if it matches itself, it does not check with itself. The only thing that may not be correct is to immediately exit the function, you probably need to, if you match yourself, then continue going through the list.

**Reply by y9Kap** [reply]
Firstly, if you are comfortable in Russian, then you can communicate in Russian, secondly, at what interval do clients enter and leave the list? I tried to do checks only by being in the list, only by ID, I tried to continue going through the list after matching = clit (by the way, a funny abbreviation), but after some time my client falls into sMQTTConnReturnIdentifierRejected and does not return from there

**Reply by terrorsl** [reply, marked as answer]
the client is added to the list after connection, and is removed when the client is disconnected or errors occur. Turn on debug messages and see what the broker writes when the client is disconnected

**Reply by y9Kap** [reply]
> the client is added to the list after connection, and is removed when the client is disconnected or errors occur. Turn on debug messages and see what the broker writes when the client is disconnected

Yes, I sat in debug mode for a very long time, but I could not understand the reason for the phantom error mqtt identified rejection

**Reply by terrorsl** [reply]
what is the number of clients and how do they behave? (how often do they disconnect, connect) more information in general. Attach log, it can help

**Reply by y9Kap** [reply]
This is exactly the problem( My client with sensors works with the broker for more than 20 hours, and then at some point it crashes with this clientID error. Only 4 clients (and only one of them publishes data, the rest just listen to subscriptions)

**Reply by terrorsl** [reply]
ok, am i right in thinking that this client connects, gives data and then disconnects and so on? if so, then look at how many clients remain when disconnecting, maybe under some circumstances it is not removed from the list.

**Reply by y9Kap** [reply]
Yes, you are right. Here is the client code: https://github.com/y9Kap/MQTTClient/blob/main/src/Client.cpp 
And yes, most likely, for some reason it remains in the list, but unfortunately, I have no way to monitor the logs throughout the day, since connection, data transfer and disconnection occur at intervals of 10 seconds, and the error appears after tens of hours. For example, yesterday everything worked fine from 22:43 in the evening until 17:38 the next day. I noticed that the data stopped transferring, opened the monitor and saw the error I wrote about above

**Reply by y9Kap** [reply]
I added TCP_FORCE_DISCONNECTED myself, because I needed to understand in which case the disconnection occurs manually, and in which the connection itself stops working

**Reply by y9Kap** [reply]
The value mqttClient.setKeepAlive(10000); is old, when I hadn't committed yet. It's been 80 for a long time

**Reply by y9Kap** [reply]
I just managed to reproduce the bug
![image](https://github.com/user-attachments/assets/18d8e5b9-d52e-4217-936b-50ede28e6142)

**Reply by y9Kap** [reply]
https://yaso.su/v0Sv0HKU
The first three connections were with MQTT_IDENTIFIER_REJECTED. Then I added another listening client and everything worked (maybe it's a coincidence, in the screenshot above there were 5 iterations without other clients and everything worked)

UPD:I was wrong, there is no error in these logs. The error appears when the second condition is met

**Reply by y9Kap** [reply]
https://yaso.su/zu4rk2sg
These logs show that when there is only one client in the list, for some reason the if block (c->getClientId() == client->getClientId()) is executed, although given that "c" is the same client as in the list--this block should not be executed

**Reply by terrorsl** [reply]
> https://yaso.su/zu4rk2sg
> 
> These logs show that when there is only one client in the list, for some reason the if block (c->getClientId() == client->getClientId()) is executed, although given that "c" is the same client as in the list--this block should not be executed
> 

What version of the broker is installed? The log shows that it writes the identifier incorrectly, this error has been fixed. I also do not see a message that the client is disconnected from the broker, but simply breaks the connection. This is bad, but not fatal. But it is worth looking at the network settings, perhaps the client has time to disconnect and connect, and the broker does not yet know about the connection break. Try to simulate the situation by turning off the power on the client and returning it back, and in the log see how quickly the message about the break arrives.

**Reply by terrorsl** [reply]
> The value mqttClient.setKeepAlive(10000); is old, when I hadn't committed yet. It's been 80 for a long time
> 
> 

Regarding the keepalive parameter, I would set the value to be less than the interval between connections, perhaps 5 seconds, so that if something goes wrong, the broker has time to kill the invalid client before the valid one connects

**Comment by y9Kap** [comment]
> What version of the broker is installed?

sMQTTBroker@0.1.8+sha.db9689d
> I also don't see messages about the client disconnecting from the broker

[3238628][D][WiFiClient.cpp:546] connected(): Disconnected: RES: 0, ERR: 128
Isn't this a disconnect message?

> The client disconnects and reconnects, and the broker doesn't know about the disconnect yet.

How can this be fixed? The client tells me that it is disconnecting.

**Reply by terrorsl** [reply]
> > I also don't see messages about the client disconnecting from the broker
> 
> [3238628][D][WiFiClient.cpp:546] connected(): Disconnected: RES: 0, ERR: 128 Isn't this a disconnect message?

No, i say about this https://github.com/terrorsl/sMQTTBroker/blob/db9689dabd4b25527bec5b87525511b384a63739/src/sMQTTClient.cpp#L50
BTW, I looked at your code and you really do forcefully terminate the connection. That's why there is no message on the broker about disconnection.
https://github.com/y9Kap/MQTTClient/blob/785592223bca428d6365348736737d6a99bb3351/src/Client.cpp#L164
https://github.com/marvinroger/async-mqtt-client/blob/89bf46485d5b60ce1e8e5e4d265a9c1570de3dc5/src/AsyncMqttClient.cpp#L707

>
>How can this be fixed? The client tells me that it is disconnecting.
https://github.com/terrorsl/sMQTTBroker/discussions/43#discussioncomment-10229246
